Maple2.Server.Game/PacketHandlers/MapleopolyHandler.cs

```diff
@@ -0,0 +1,191 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Maple2.Database.Storage;
+using Maple2.Model.Enum;
+using Maple2.Model.Error;
+using Maple2.Model.Game;
+using Maple2.Model.Game.Event;
+using Maple2.Model.Metadata;
+using Maple2.PacketLib.Tools;
+using Maple2.Server.Core.Constants;
+using Maple2.Server.Core.PacketHandlers;
+using Maple2.Server.Game.Packets;
+using Maple2.Server.Game.Session;
+
+namespace Maple2.Server.Game.PacketHandlers;
+
+public class MapleopolyHandler : PacketHandler<GameSession> {
+    public override RecvOp OpCode => RecvOp.Mapleopoly;
+
+    #region Autofac Autowired
+    // ReSharper disable MemberCanBePrivate.Global
+    public required ItemMetadataStorage ItemMetadata { private get; init; }
+    public required GameStorage GameStorage { private get; init; }
+    // ReSharper restore All
+    #endregion
+
+    private enum Command : byte {
+        Load = 0,
+        Roll = 1,
+        Result = 3,
+    }
+
+    public override void Handle(GameSession session, IByteReader packet) {
+        var function = packet.Read<Command>();
+        switch (function) {
+            case Command.Load:
+                HandleLoad(session);
+                return;
+            case Command.Roll:
+                HandleRoll(session);
+                return;
+            case Command.Result:
+                HandleResult(session);
+                return;
+        }
+    }
+
+    private void HandleLoad(GameSession session) {
+        using GameStorage.Request db = GameStorage.Context();
```

Comment:
Is it necessary to read the event from db every time? I assume the same event is used across all players on the server. It might make more sense to manage this centrally (e.g. `GameServer`, and retrieve it from memory)

Rough idea of what this would look like:

`GameServer.cs`
```cs
public T? FindEvent<T>() where T : GameEvent {
    if (!eventCache.TryGetValue(typeof(T).Name, out GameEvent? gameEvent) {
        gameEvent = ...; // Lookup from db
        eventCache[typeof(T).Name] = gameEvent;
    }

    // You can check expiry here as well if you want.
    return gameEvent as T;
}
```

`GameSession.cs`
```cs
public T? FindEvent<T>() where T : GameEvent {
    return server.FindEvent<T>();
}
```

`MapleopolyHandler.cs`
```cs
BlueMarble? blueMarble = session.FindEvent<BlueMarble>();
if (blueMarble == null {
    return;
}
...
```

Reply:
I see what you mean. I would need to change `FindEvent` (in both GameServer and GameSession) to something of the following:

```csharp    
public GameEvent? FindEvent<T>() where T : GameEventInfo {
...
        return gameEvent;
    }
```

considering I need to search by GameEventInfo but get GameEvent returned instead

However would it make sense to null suppress in MapleopolyHandler?

```csharp
        GameEvent? gameEvent = session.FindEvent<BlueMarble>();
        if (gameEvent == null) {
            return;
        }
        BlueMarble blueMarble = (gameEvent.EventInfo as BlueMarble)!;
        ```

Considering this was already checked if it was BlueMarble with both FindEvent functions, it should be safe to assume it _will_ be BlueMarble, no?

Reply:
Oh right, didn't realize it was `GameEvent` vs `GameEventInfo`. I think what you have is reasonable, I had an idea around making `GameEvent` into `GameEvent<T> where T : GameEventInfo`, but it gets complicated and I'm not sure if it even works.

To access db, just add `GameStorage gameStorage` to the constructor and it will be injected by autofac.